### ISSUE TITLE
HttpContext RequestAborted no longer works.

### ISSUE THREAD

**Issue by blogcraft** [dotnet/aspnetcore#38917]
### Describe the bug
Since .NET 6, the `HttpContext.RequestAborted` does not updates when the request is cancelled. This used to work with .NET 5.
IsCancellationRequested stays in false even tho the request has been canceled and CanBeCancelled is set to true.

### To Reproduce
I started a .NET 6 Angular project and created an HTTP GET request and subscribe to it, the Request is in an infinite loop intentionally so it can only be ended by the cancellation token. Also have a subject to cancel the request with an other button.

In .NET 5 this works flawlessly. In .NET 6 its unusable.

An Example Repo can be found here: https://github.com/blogcraft/TestCancel

### Further technical details
- ASP.NET Core version: .NET 6.0.100
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version: VS Code
- Include the output of `dotnet --info`:

<details>
  <summary>SDK de .NET (que refleje cualquier global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Entorno de tiempo de ejecución:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100\

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa

.NET SDKs installed:
  3.1.415 [C:\Program Files\dotnet\sdk]
  5.0.401 [C:\Program Files\dotnet\sdk]
  6.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download</summary>
</details>

**Comment by davidfowl** [comment]
cc @stephentoub, just in case there's any insight here.

**Comment by stephentoub** [comment]
> just in case there's any insight here.

Haven't tried the repro, but if token.CanBeCanceled is true and token.IsCancellationRequested is false, it simply sounds like no one is canceling the CancellationTokenSource.

**Comment by martincostello** [comment]
Could it be maybe something to do with the `CancellationTokenSource` pooling/re-use that was added in 6.0?

**Comment by davidfowl** [comment]
> Could it be maybe something to do with the CancellationTokenSource pooling/re-use that was added in 6.0?

Yes it could be. The change was [very small](https://github.com/dotnet/aspnetcore/commit/8d721cf5fb3d65b65042f9f670e6669173ee6416#diff-5ae4f2936177430c438bd4d4f619b0b4342879626806614d2688bac5eff3e564) but I haven't spent any time investigating it to see if there's a bug somewhere.

**Comment by davidfowl** [comment]
@blogcraft can you verify the HTTP version you are using? Is it HTTP/2 or HTTP/1.1

**Comment by blogcraft** [comment]
@davidfowl Can You help me with that? My development environment I use the project _as is_ (just the same as I used with .NET 5). Is there a way to force HTTP/2?
In my production environment I use an Azure AppService that has always been configured tu use HTTP 2.0.

**Comment by Tratcher** [comment]
@javiercn I think this is because of the new SPA proxy in 6.0. Can you take a look?

**Comment by javiercn** [comment]
@Tratcher from what I can tell, the "cancellation" is not even making it back to the proxy.

I added some logs to the proxy code to validate this.
```
const PROXY_CONFIG = [
  {
    context: [
      "/weatherforecast",
   ],
    target: target,
    secure: false,
    onProxyRes: (proxyRes, req, res) => {
      console.log(res.statusCode);
      console.log(proxyRes.statusCode);
    },
  }
]
```

I see a bunch of `ok...` repeated in the console, so it seems something is stuck in a loop.

Other requests come back normally:
![image](https://user-images.githubusercontent.com/6995051/154091077-d477a780-2353-40d9-8031-c7a9332b9aee.png)

But the cancellations seem to loop and don't send an actual cancellation/failure upstream.
![image](https://user-images.githubusercontent.com/6995051/154091331-80292395-2e25-4ec0-98bb-c20e07cab9b0.png)

It seems from the network trace that the request is pending:
![image](https://user-images.githubusercontent.com/6995051/154091638-740f397d-d079-4bda-a545-37f7b8312b0d.png)

One way to go about this, is to add a timeout on the proxy code in your proxy.conf.js for example `    proxyTimeout: 3000,`. 
That should make sure that the proxy sends back a response instead of holding indefinitely when it doesn't receive a response from the server.

You can check how to fine-tune this on https://github.com/chimurai/http-proxy-middleware#readme

**Comment by adityamandaleeka** [comment]
@javiercn Can you please share more about what the problem is here? It sounds like a SPA proxy issue, but I see you've moved it to runtime so I want to check what we need to do.

**Comment by javiercn** [comment]
Yes.

The proxy is not seeing any response from the server. I gave the customer an option but I want us to discard that there is no underlying issue at the Kestrel/proxy level.

When kestrel aborts the request is it supposed to notify the upstream connection/server in any way?


Best regards,

Javier Calvarro Nelson

________________________________

**Comment by Tratcher** [comment]
> When kestrel aborts the request is it supposed to notify the upstream connection/server in any way?

Upstream server? You mean like client -> kestrel -> reverse proxy -> upstream? In that case Kestrel's RequestAborted CancellationToken is the signal through the app.

If you meant client -> upstream -> reverse proxy -> Kestrel, then a request abort is signaled by closing the associated TCP connection or HTTP/2 stream.

**Comment by davidfowl** [comment]
We need look at the logic here https://github.com/http-party/node-http-proxy and see if there's an issue on their end with propagating client disconnection maybe.

**Comment by javiercn** [comment]
@davidfowl I think setting the proxyTimeout on our templates is a good solution while we determine what is going on.

**Comment by singh733** [comment]
@javiercn  When we click on the Cancel Request Button , that call never hit the below breakpoint. 
If this is the First method which accepts the request then cancel request never to the asp.net core.
Please let me know is this the first method which accepts the Request?

![image](https://user-images.githubusercontent.com/83835739/168020066-15d5fbce-c068-4b0a-9aee-2f88889140ad.png)

**Comment by javiercn** [comment]
@singh733 not sure what you are talking about. If you are facing a problem, please file a separate issue and provide the required context there. If your problem is similar to the one described here, you can try using the solution we suggested, which is to add `proxyTimeout` to the proxy configuration.

**Comment by singh733** [comment]
@javiercn I was trying to resolve the above issue. I don't think it is asp.net core problem because the cancel button calls does not hit the EXECUTE method of  HTTP2STREAM

**Comment by davidfowl** [comment]
@singh733 you're looking at the wrong method, that's not how this works. I think you're a bit in the weeds here. My guess is that the nodejs proxy (the client) isn't disconnecting properly and that's likely why the client isn't seeing the disconnect.

**Comment by Lonli-Lokli** [comment]
- HTTP/2 uses RST_STREAM frame on cancel
- HTTP/3 uses QUIK RESET_STREAM frame

**Comment by patrickconcannon** [comment]
Hi @davidfowl, has there been any solution to this issue?

**Comment by veantm** [comment]
This issue seems to be related to https://github.com/nodejs/node/issues/46666
